file.go

```diff
@@ -7,56 +7,97 @@ import (
+	// all files stored in the tree should be returned). Some implementations may
+	// choose not to implement this
+	Size() (int64, error)
+}
+
+// Node represents a regular Unix file
 type File interface {
-	// Files implement ReadCloser, but can only be read from or closed if
-	// they are not directories
-	io.ReadCloser
-
-	// FileName returns a filename associated with this file
-	FileName() string
-
-	// FullPath returns the full path used when adding with this file
-	FullPath() string
-
-	// IsDirectory returns true if the File is a directory (and therefore
-	// supports calling `NextFile`) and false if the File is a normal file
-	// (and therefor supports calling `Read` and `Close`)
-	IsDirectory() bool
-
-	// NextFile returns the next child file available (if the File is a
-	// directory). It will return (nil, io.EOF) if no more files are
-	// available. If the file is a regular file (not a directory), NextFile
-	// will return a non-nil error.
-	NextFile() (File, error)
+	Node
+
+	io.Reader
+	io.Seeker
 }
 
-type StatFile interface {
-	File
+// DirEntry exposes information about a directory entry
+type DirEntry interface {
+	// Name returns base name of this entry, which is the base name of referenced
+	// file
+	Name() string
 
-	Stat() os.FileInfo
+	// Node returns the file referenced by this DirEntry
+	Node() Node
+
+	// File is an alias for ent.Node().(File). If the file isn't a regular
+	// file, nil value will be returned
+	File() File
+
+	// Dir is an alias for ent.Node().(directory). If the file isn't a directory,
+	// nil value will be returned
+	Dir() Directory
 }
 
-type PeekFile interface {
-	SizeFile
+// DirIterator is a iterator over directory entries.
+// See Directory.Entries for more
+type DirIterator interface {
+	// DirEntry holds information about current directory entry.
+	// Note that after creating new iterator you MUST call Next() at least once
+	// before accessing these methods. Calling these methods without prior calls
+	// to Next() and after Next() returned false may result in undefined behavior
+	DirEntry
+
+	// Next advances iterator to the next file.
+	Next() bool
 
-	Peek(n int) File
-	Length() int
+	// Err may return an error after previous call to Next() returned `false`.
+	// If previous call to Next() returned `true`, Err() is guaranteed to
+	// return nil
+	Err() error
 }
 
-type SizeFile interface {
-	File
+// Directory is a special file which can link to any number of files.
+type Directory interface {
+	Node
 
-	Size() (int64, error)
+	// Entries returns a stateful iterator over directory entries.
+	//
+	// Example usage:
+	//
+	// it := dir.Entries()
+	// for it.Next() {
+	//   name := it.Name()
+	//   file := it.Node()
+	//   [...]
+	// }
+	// if it.Err() != nil {
+	//   return err
+	// }
+	//
+	// Note:
+	// - Some implementations of this functions may define some constraints in how
+	//   it can be used
+	Entries() (DirIterator, error)
```

Comment:
Note to self: make this not return `error` since it can be done with iterator anyways